core/jobs/types/model_property.py

```diff
@@ -109,62 +128,95 @@ def yield_value_from_model(self, model):
         else:
             yield value
 
-    def _to_model_class(self):
+    def _to_model_class(self) -> Type[base_models.BaseModel]:
         """Returns the model class associated with this instance.
 
         Returns:
             type(BaseModel). The model type.
         """
-        return job_utils.get_model_class(self._model_kind)
-
-    def _to_property(self):
+        # Here, get_model_class() returns the value of type
+        # Type[datastore_services.Model] but the expected return value from
+        # this method is Type[base_models.BaseModel]. So, that's why we used
+        # cast here.
+        return cast(
+            Type[base_models.BaseModel],
+            job_utils.get_model_class(self._model_kind)
+        )
+
+    def _to_property(
+        self
+    ) -> Union[
+        datastore_services.Property,
+        Callable[[base_models.BaseModel], str]
+    ]:
         """Returns the Property object associated with this instance.
 
         Returns:
             *. A property instance.
         """
-        return getattr(self._to_model_class(), self._property_name)
-
-    def _is_repeated_property(self):
+        # Using a cast here because the return value of getattr() method is
+        # dynamic and mypy will assume it to be Any type otherwise.
+        return cast(
+            Union[
+                datastore_services.Property,
+                Callable[[base_models.BaseModel], str]
+            ],
+            getattr(self._to_model_class(), self._property_name)
+        )
```

Comment:
ditto as above, but I guess here it might be harder?

Reply:
I think casting a Any value is better than asserting it to a specific type. because assert is checked during runtime and cast is not. So if the assert is used with Any then there might be chances that it fails ( maybe, in the future ).

(In my opinion) assert should be used when we are sure about some information on the type.

Reply:
Here, I think assert is not applicable because:
if we do ```assert isinstance(getattr(self._to_model_class(), self._property_name), Union[
                datastore_services.Property,Callable[[base_models.BaseModel], str]])``` then MyPy gives error. because internally `Callable[]` is implemented as object not class, but `isinstance()` always expects a class to compare with. ( refrence: https://github.com/python/cpython/blob/9478b263a3279ebd445d722b1fdcdd1575bed652/Lib/typing.py#L2625 )

I also tried `assert isinstance(getattr(self._to_model_class(), self._property_name), datastore_services.Property) and callable(getattr(self._to_model_class(), self._property_name))`, this fixes the MyPy test but fails backend test. Because during run time python property is considered as normal attribute not function, but during type checking it is considered as `callable` because it is decorate using python's property class. 

This method ( _to_property() ) is a protected method and it accepts no extra arguments instead it uses the instance variables (self._property_name) which are already validated in the constructor. So there are high chances that the return value of `getattr()` is within the `Union[datastore_services.Property,Callable[[base_models.BaseModel], str]]` type range.

Please correct me if I'm wrong !!!
Thanks!

Reply:
Thanks! Great explanation.

Can the assert checks be implemented at least partially? So that the `cast` is at least partially verified?

Reply:
I tried a new approach PTAL!

Basically, I split assertions into 2 parts using `if MYPY:` clause. During type checking `Callable[]` assertion is called while during runtime instance assertion is invoked.